### PR TITLE
Remove nose imports

### PR DESCRIPTION
**Pull request by Adersh97**

### CODE REVIEW

**Review by twizmwazin — COMMENTED** [review]
So typically when using unittest's asserts, it is in methods of existing `TestCase` objects. For tests that are standalone functions, we either want to convert them to be methods of `TestCase` subclasses, or probably use the `assert` keyword instead.

File: tests/analyses/reaching_definitions/test_subject.py [hunk 1]
```diff
@@ -19,8 +17,8 @@ def test_can_be_instantiated_with_a_function(_):
     function = _a_mock_function(0x42, 'function_name')
     subject = Subject(function)
 
-    nose.tools.assert_equals(subject.content, function)
-    nose.tools.assert_equals(subject.type, SubjectType.Function)
+    assert (subject.content == function)
+    assert (subject.type == SubjectType.Function)
```
Comment (ltfish) [inline tests/analyses/reaching_definitions/test_subject.py:21]: Remove the surrounding parentheses please. It should be `assert subject.type == SubjectType.Function`.

File: tests/analyses/reaching_definitions/test_subject.py [hunk 2]
```diff
@@ -29,20 +27,25 @@ def test_can_be_instantiated_with_a_block(_):
     block = Block(0x42, byte_string=b'', arch=arch)
     subject = Subject(block)
 
-    nose.tools.assert_equals(subject.content, block)
-    nose.tools.assert_equals(subject.type, SubjectType.Block)
+    assert subject.content == block
+    assert (subject.type == SubjectType.Block)
```
Comment (ltfish) [inline tests/analyses/reaching_definitions/test_subject.py:31]: Same as above.

File: tests/analyses/reaching_definitions/test_subject.py [hunk 3]
```diff
@@ -54,23 +57,30 @@ def test_when_instanciated_with_a_function_need_other_attributes(_, __):
...
     block = Block(0x42, byte_string=b'', arch=arch)
     subject = Subject(block)
 
-    with nose.tools.assert_raises(TypeError):
+    flag = False
+    try:
         _ = subject.func_graph
+    except TypeError:
+        flag = True
+    assert flag
```
Comment (ltfish) [inline tests/analyses/reaching_definitions/test_subject.py:86]: Let's not reimplement `assertRaises`. Can you rewrite this function to use `unittest.TestCase` so you can call `self.assertRaises()` there?

File: tests/test_argv.py [hunk 4]
```diff
@@ -29,20 +28,20 @@ def test_mips():
     found = xpl.found[0]
     conc = found.solver.eval(found.memory.load(found.registers.load('sp'), 400), cast_to=bytes)
 
-    nose.tools.assert_equal(b"Yan is a noob" in conc, True)
+    assert b"Yan is a noob" in conc == True
```
Comment (ltfish) [inline tests/test_argv.py:31]: `== True` is redundant. It should be `assert b"Yan is a noob" in conc`.

**Review by ltfish — CHANGES_REQUESTED** [review]
I pointed out several prevalent mistakes. Please go through your PR and fix these issues accordingly. Thanks!

File: tests/test_stack_alignment.py [hunk 5]
```diff
@@ -25,8 +24,7 @@ def test_alignment():
             cc.setup_callsite(st, 0, [0x1337])
 
             # ensure stack alignment is correct
-            nose.tools.assert_true(st.solver.is_true(((st.regs.sp + cc.STACKARG_SP_DIFF) % cc.STACK_ALIGNMENT == 0)),
-                                   'non-zero stack alignment after setup_callsite for %s'%cc)
+            assert st.solver.is_true(((st.regs.sp + cc.STACKARG_SP_DIFF) % cc.STACK_ALIGNMENT == 0))
```
Comment (twizmwazin) [inline tests/test_stack_alignment.py:27]: Please preserve the message

File: tests/test_state.py [hunk 6]
```diff
@@ -42,73 +40,72 @@ def test_state_merge():
...
-    nose.tools.assert_true(c.has_plugin('libc'))
-    nose.tools.assert_true(d.has_plugin('libc'))
+    assert c.has_plugin('libc')
+    assert d.has_plugin('libc')
 
     # test merging posix with different open files (illegal!)
     a = SimState(arch='AMD64', mode='symbolic')
     b = a.copy()
     a.posix.open(b'/tmp/idk', 1)
-    nose.tools.assert_raises(angr.errors.SimMergeError, lambda: a.copy().merge(b.copy()))
```
Comment (twizmwazin) [inline tests/test_state.py:111]: Looks like this got dropped without a replacement?

File: tests/test_state.py [hunk 7]
```diff
@@ -262,9 +259,8 @@ def test_successors_catch_arbitrary_interrupts():
 
     simgr.step()
 
-    nose.tools.assert_equal(len(simgr.errored), 0, msg="The state should not go to the errored stash. Is "
-                                                       "AngrSyscallError handled in SimSuccessors?")
-    nose.tools.assert_equal(len(simgr.unsat), 1)
+    assert len(simgr.errored) == 0
```
Comment (twizmwazin) [inline tests/test_state.py:262]: Please preserve the messages

File: tests/test_tracer.py [hunk 8]
```diff
@@ -113,26 +112,23 @@ def test_cgc_se1_palindrome_raw():
...
-    nose.tools.assert_false(simgr.crashed)
+    assert not simgr.crashed
 
     # make sure angr modeled the correct output
     stdout_dump = simgr.traced[0].posix.dumps(1)
-    nose.tools.assert_true(stdout_dump.startswith(b"\nWelcome to Palindrome Finder\n\n"
-                                                  b"\tPlease enter a possible palindrome: "
-                                                  b"\t\tYes, that's a palindrome!\n\n"
-                                                  b"\tPlease enter a possible palindrome: "))
+    assert stdout_dump.startswith(b"\nWelcome to Palindrome Finder\n\n"                                                   b"\tPlease enter a possible palindrome: "                                                   b"\t\tYes, that's a palindrome!\n\n"                                                   b"\tPlease enter a possible palindrome: ")
```
Comment (twizmwazin) [inline tests/test_tracer.py:123]: Formatting

File: tests/test_types.py [hunk 9]
```diff
@@ -111,159 +111,158 @@ def test_struct_deduplication():
...
     fdef = angr.types.parse_defns("int f(int param_1, int param_2);") # type: Dict[str, SimTypeFunction]
     sig = fdef['f']
-    nose.tools.assert_equal(sig.arg_names, ['param_1', 'param_2'])
+    assert sig.arg_names == ['param_1', 'param_2']
 
     # Check that arg_names survive a with_arch call
     nsig = sig.with_arch(angr.archinfo.ArchAMD64())
-    nose.tools.assert_equal(sig.arg_names, nsig.arg_names,
-                            "Function type generated with .with_arch() doesn't have identical arg_names")
+    assert sig.arg_names == nsig.arg_names
```
Comment (twizmwazin) [inline tests/test_types.py:222]: Please preserve the message

File: tests/test_unicorn.py [hunk 10]
```diff
@@ -119,11 +118,7 @@ def broken_fauxware_arm():
     pg = p.factory.simulation_manager(s_unicorn)
     pg.explore()
     assert all("Unicorn" in ''.join(p.history.descriptions.hardcopy) for p in pg.deadended)
-    nose.tools.assert_equal(sorted(pg.mp_deadended.posix.dumps(1).mp_items), sorted((
-        b'Username: \nPassword: \nWelcome to the admin console, trusted user!\n',
-        b'Username: \nPassword: \nGo away!',
-        b'Username: \nPassword: \nWelcome to the admin console, trusted user!\n'
-    )))
+    assert sorted(pg.mp_deadended.posix.dumps(1).mp_items) == sorted((         b'Username: \nPassword: \nWelcome to the admin console, trusted user!\n',         b'Username: \nPassword: \nGo away!',         b'Username: \nPassword: \nWelcome to the admin console, trusted user!\n'     ))
```
Comment (twizmwazin) [inline tests/test_unicorn.py:121]: Formatting

File: tests/test_variablerecovery.py [hunk 11]
```diff
@@ -106,9 +105,7 @@ def run_variable_recovery_analysis(func_name, groundtruth, is_fast):
                 l.error('Unsupported variable sort %s.', var_sort)
                 assert False
 
-            nose.tools.assert_is_not_none(the_var, msg="The variable %s in groundtruth at instruction %#x cannot be "
-                                                       "found in variable manager." % (var_info, insn_addr)
-                                          )
+            assert the_var is not None
```
Comment (twizmwazin) [inline tests/test_variablerecovery.py:108]: Please preserve the message

File: tests/test_vex.py [hunk 12]
```diff
@@ -443,11 +441,8 @@ def test_store_simplification():
     sim_successors = HeavyVEXMixin(None).process(state.copy(), irsb=irsb)
     exit_state = sim_successors.all_successors[0]
 
-    nose.tools.assert_true(claripy.backends.z3.is_true(exit_state.regs.ebp == state.regs.esp - 4))
-
-
```
Comment (twizmwazin) [inline tests/test_vex.py:448]: Looks like a few too many newlines were removed here

File: tests/test_vfg.py [hunk 13]
```diff
@@ -41,18 +40,14 @@ def run_vfg_buffer_overflow(arch):
...
     states = vfg.final_states
-    nose.tools.assert_equal(len(states), 2)
+    assert len(states) == 2
     stack_check_fail = proj._extern_obj.get_pseudo_addr('symbol hook: __stack_chk_fail')
-    nose.tools.assert_equal(set([ s.solver.eval_one(s.ip) for s in states ]),
-                            {
-                                stack_check_fail,
-                                0x4005b4
-                            })
+    assert set([ s.solver.eval_one(s.ip) for s in states ]) == {                                 stack_check_fail,                                 0x4005b4                             }
```
Comment (twizmwazin) [inline tests/test_vfg.py:47]: Formatting

File: tests/test_windows_stack_cookie.py [hunk 14]
```diff
@@ -45,7 +44,6 @@ def test_security_cookie_init():
     check_value(project, angr.simos.windows.SecurityCookieInit.STATIC, compare_static)
     check_value(project, angr.simos.windows.SecurityCookieInit.SYMBOLIC, compare_symbolic)
 
-    nose.tools.assert_raises(TypeError, project.factory.blank_state, security_cookie_init=1)
```
Comment (twizmwazin) [inline tests/test_windows_stack_cookie.py:48]: This should be converted to unittest to use its equivalent.

**Review by twizmwazin — CHANGES_REQUESTED** [review]
I left a bunch of inline comments for various issues. I tried to only leave one comment per type of issue per file, so please check if a problem exists multiple times within that file. Thank you!

**Review by twizmwazin — CHANGES_REQUESTED** [review]
Great progress! Please address the inline comments plus the CI failures.

File: tests/test_cfg_thumb_firmware.py [hunk 15]
```diff
@@ -25,16 +24,16 @@ def test_thumb_firmware_cfg():
...
-    assert_true(len(block.successors()) == 1)
+    assert len(block.successors()) == 1
     succ = list(block.successors())[0]
-    assert_true(succ.addr == 0x080081dd)
+    assert succ.addr == 0x080081dd
     f2 = p.kb.functions[succ.addr]
-    assert_true(f2.name == '_vfprintf_r')
-    assert_true(f2.returning)
+    assert f2.name == '_vfprintf_r'
+    assert f2.returning
```
Comment (twizmwazin) [inline tests/test_cfg_thumb_firmware.py:36]: trailing whitespace

File: tests/test_keystone.py [hunk 16]
```diff
@@ -18,6 +20,8 @@
     'mips': b"addi $1, $1, 0xf"
 }
 
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_keystone.py:23]: Looks like you missed a git conflict

File: tests/test_scanf.py [hunk 17]
```diff
@@ -90,6 +93,114 @@ def check(self, path):
         return check_passes
 
 
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_scanf.py:96]: Another git conflict

File: tests/test_spiller.py [hunk 18]
```diff
@@ -1,28 +1,74 @@
...
 def pickle_callback(state):
     state.globals['pickled'] = True
+
+
 def unpickle_callback(sid, state):  # pylint:disable=unused-argument
     state.globals['unpickled'] = True
+
+
 def priority_key(state):
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_spiller.py:21]: git conflict

File: tests/test_sscanf.py [hunk 19]
```diff
@@ -1,51 +1,61 @@
...
-        raise nose.SkipTest()
+class TestSscanf(unittest.TestCase):
+    def test_sscanf(self):
+        if not sys.platform.startswith('linux'):
+            return
+
+        test_bin = os.path.join(test_location, "..", "..", "binaries", "tests", "x86_64", "sscanf_test")
+        b = angr.Project(test_bin)
 
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_sscanf.py:24]: git conflict

File: tests/test_unicorn.py [hunk 20]
```diff
@@ -17,183 +15,350 @@ def _remove_addr_from_trace_item(trace_item_str):
...
             # we do not care if addresses of SimProcedures match, since they are not allocated in a deterministic way
             trace_item_str = _remove_addr_from_trace_item(trace_item_str)
             expected_str = _remove_addr_from_trace_item(expected_str)
 
-        nose.tools.assert_equal(trace_item_str, expected_str)
+        assert trace_item_str == expected_str
+
 
 def test_stops():
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_unicorn.py:34]: git conflicts, quite a few in this file

File: tests/test_vault.py [hunk 21]
```diff
@@ -1,113 +1,153 @@
...
-	# yield do_ast_vault, angr.vaults.VaultDirShelf
+    yield do_ast_vault, angr.vaults.VaultDir
+    yield do_ast_vault, angr.vaults.VaultShelf
+    yield do_ast_vault, angr.vaults.VaultDict
+    # VaultDirShelf does not guarantee identity equivalence due to the absence of caching
+    # yield do_ast_vault, angr.vaults.VaultDirShelf
+
 
 def test_project():
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_vault.py:117]: git conflict

File: tests/test_vfg.py [hunk 22]
```diff
@@ -3,24 +3,32 @@
...
+)
+
+vfg_buffer_overflow_addresses = {"x86_64": 0x40055C}
 
-vfg_buffer_overflow_addresses = {
-    'x86_64': 0x40055c
-}
 
 def run_vfg_buffer_overflow(arch):
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_vfg.py:20]: git conflict

File: tests/test_windows_stack_cookie.py [hunk 23]
```diff
@@ -38,15 +37,25 @@ def check_value(project, init_type, comparison):
     comparison(state, value)
 
 
+<<<<<<< HEAD
```
Comment (twizmwazin) [inline tests/test_windows_stack_cookie.py:40]: git conflict

**Review by twizmwazin — COMMENTED** [review]
Please be careful to make sure you resolve all of the git conflicts! In the future, we'll probably want to break massive PRs like this down in to more manageable chunks that can be merged more quickly to avoid conflicts.

File: tests/test_keystone.py [hunk 24]
```diff
@@ -18,45 +20,49 @@
...
 
-    for arch_name in insn_texts:
-        yield run_keystone, arch_name
+        # Installing keystone on Windows is currently a pain. Fix the installation first (may it pip installable) before
+        # re-enabling this test on Windows.
+        if not sys.platform.startswith('linux'):
+            return
+
+        for arch_name in insn_texts:
+            yield run_keystone, arch_name
```
Comment (twizmwazin) [inline tests/test_keystone.py:62]: I believe this should be `self.run_keystone`, and that is what is causing the weird CI failure

**Review by twizmwazin — APPROVED** [review]
LGTM once the CI passes

**Review by twizmwazin — CHANGES_REQUESTED** [review]
I went through the past comments to double check that the issues were addressed, a few were missed. Please mark the comments resolved once they are fixed, and then this will LGTM.

**Comment by twizmwazin** [comment]
Thanks!

**Pull request merged**